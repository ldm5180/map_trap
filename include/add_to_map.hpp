#ifndef ADD_TO_MAP_HPP
#define ADD_TO_MAP_HPP

#include <algorithm>
#include <benchmark/benchmark.h>
#include <cstdlib>
#include <string>

static const auto add_to_map = [](const auto num_elems, auto &map) {
  // Add elements to map, this sorts automatically.
  for (auto i = 0; i < num_elems; ++i) {
    const auto random_key = std::rand();
    map[random_key] = std::to_string(map.size());
  }
  benchmark::ClobberMemory();
};

static const auto add_to_map_and_key_list = [](const auto num_elems, auto &map,
                                               auto &keys) {
  // Add elements to map, this sorts automatically.
  for (auto i = 0; i < num_elems; ++i) {
    const auto random_key = std::rand();
    map[random_key] = std::to_string(map.size());
    keys.push_back(random_key);
    std::sort(keys.begin(), keys.end());
  }
  benchmark::ClobberMemory();
};

#endif // ADD_TO_MAP_HPP
