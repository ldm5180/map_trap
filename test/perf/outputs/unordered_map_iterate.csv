Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:39:14
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_unordered_map_iterate/8",381737,1830.24,1822.16,ns,,,,,
"BM_unordered_map_iterate/16",81484,8625.49,8625.07,ns,,,,,
"BM_unordered_map_iterate/32",14539,48788.1,48786.6,ns,,,,,
"BM_unordered_map_iterate/64",1989,326207,325582,ns,,,,,
"BM_unordered_map_iterate/128",445,1.53505e+06,1.53311e+06,ns,,,,,
