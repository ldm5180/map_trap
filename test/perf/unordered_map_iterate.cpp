#include "add_to_map.hpp"

#include <benchmark/benchmark.h>
#include <cstdlib>
#include <string>
#include <unordered_map>

using map_t = std::unordered_map<int, std::string>;

static void BM_unordered_map_iterate(benchmark::State &state) {
  std::srand(42);
  map_t map{};
  map.reserve(state.range(0));
  std::vector<map_t::key_type> keys{};
  keys.reserve(state.range(0));

  for (auto i = 0; i < state.range(0); ++i) {
    add_to_map(state.range(0), map);
  }

  for (auto _ : state) {
    keys.clear();
    std::transform(map.begin(), map.end(), std::back_inserter(keys),
                   [](const auto &kv) { return kv.first; });
    std::sort(keys.begin(), keys.end());
    map_t::mapped_type v{};
    for (const auto &k : keys) {
      benchmark::DoNotOptimize(v = map[k]);
    }
  }
}
BENCHMARK(BM_unordered_map_iterate)->RangeMultiplier(2)->Range(8, 8 << 4);

BENCHMARK_MAIN();
