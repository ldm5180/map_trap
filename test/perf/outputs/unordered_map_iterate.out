Run on (4 X 2594.09 MHz CPU s)
2017-10-28 16:27:38
name,iterations,real_time,cpu_time,time_unit,bytes_per_second,items_per_second,label,error_occurred,error_message
"BM_unordered_map_iterate/8",375725,1837.76,1827.07,ns,,,,,
"BM_unordered_map_iterate/16",85494,8486.71,8490.75,ns,,,,,
"BM_unordered_map_iterate/32",14080,45127.5,45149.7,ns,,,,,
"BM_unordered_map_iterate/64",2262,307893,308023,ns,,,,,
"BM_unordered_map_iterate/128",413,1.49564e+06,1.49637e+06,ns,,,,,
